packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts

```diff
@@ -97,92 +97,20 @@ export type CustomStyleHooksContextValue = {
-  usePresenceBadgeStyles_unstable: noop,
-  useButtonStyles_unstable: noop,
-  useCompoundButtonStyles_unstable: noop,
-  useMenuButtonStyles_unstable: noop,
-  useSplitButtonStyles_unstable: noop,
-  useToggleButtonStyles_unstable: noop,
-  useCheckboxStyles_unstable: noop,
-  useComboboxStyles_unstable: noop,
-  useDropdownStyles_unstable: noop,
-  useListboxStyles_unstable: noop,
-  useOptionStyles_unstable: noop,
-  useOptionGroupStyles_unstable: noop,
-  useDividerStyles_unstable: noop,
-  useInputStyles_unstable: noop,
-  useImageStyles_unstable: noop,
-  useLabelStyles_unstable: noop,
-  useLinkStyles_unstable: noop,
-  useMenuDividerStyles_unstable: noop,
-  useMenuGroupHeaderStyles_unstable: noop,
-  useMenuGroupStyles_unstable: noop,
-  useMenuItemCheckboxStyles_unstable: noop,
-  useMenuItemRadioStyles_unstable: noop,
-  useMenuItemStyles_unstable: noop,
-  useMenuListStyles_unstable: noop,
-  useMenuPopoverStyles_unstable: noop,
-  useMenuSplitGroupStyles_unstable: noop,
-  usePersonaStyles_unstable: noop,
-  usePopoverSurfaceStyles_unstable: noop,
-  useRadioGroupStyles_unstable: noop,
-  useRadioStyles_unstable: noop,
-  useSelectStyles_unstable: noop,
-  useSliderStyles_unstable: noop,
-  useSpinButtonStyles_unstable: noop,
-  useSpinnerStyles_unstable: noop,
-  useSwitchStyles_unstable: noop,
-  useTabStyles_unstable: noop,
-  useTabListStyles_unstable: noop,
-  useTextStyles_unstable: noop,
-  useTextareaStyles_unstable: noop,
-  useTooltipStyles_unstable: noop,
-  useDialogTitleStyles_unstable: noop,
-  useDialogBodyStyles_unstable: noop,
-  useDialogActionsStyles_unstable: noop,
-  useDialogSurfaceStyles_unstable: noop,
-  useDialogContentStyles_unstable: noop,
-  useProgressBarStyles_unstable: noop,
-  useToolbarButtonStyles_unstable: noop,
-  useToolbarRadioButtonStyles_unstable: noop,
-  useToolbarGroupStyles_unstable: noop,
-  useToolbarToggleButtonStyles_unstable: noop,
-  useToolbarDividerStyles_unstable: noop,
-  useToolbarStyles_unstable: noop,
-  useTableCellStyles_unstable: noop,
-  useTableRowStyles_unstable: noop,
-  useTableBodyStyles_unstable: noop,
-  useTableStyles_unstable: noop,
-  useTableHeaderStyles_unstable: noop,
-  useTableHeaderCellStyles_unstable: noop,
-  useTableResizeHandleStyles_unstable: noop,
-  useTableSelectionCellStyles_unstable: noop,
-  useTableCellActionsStyles_unstable: noop,
-  useTableCellLayoutStyles_unstable: noop,
-  useDataGridCellStyles_unstable: noop,
-  useDataGridRowStyles_unstable: noop,
-  useDataGridBodyStyles_unstable: noop,
-  useDataGridStyles_unstable: noop,
-  useDataGridHeaderStyles_unstable: noop,
-  useDataGridHeaderCellStyles_unstable: noop,
-  useDataGridSelectionCellStyles_unstable: noop,
+
+const preProxyDefaultValue: Partial<CustomStyleHooksContextValue> = {};
+
+const defaultValueProxyHandler: ProxyHandler<CustomStyleHooksContextValue> = {
+  get(_target, _prop, _receiver) {
+    return noop;
+  },
 };
 
+const customStyleHooksContextDefaultValue = new Proxy(
```

Comment:
@GeoffCoxMSFT I don't think that it works as expected as after a shallow merge in `FluentProvider` we are losing the proxy functionality. An example is below: 

```js
const noop = () => { };
const preProxyDefaultValue = {};

const defaultValueProxyHandler = {
    get() {
        return noop;
    },
};

const customStyleHooksContextDefaultValue = new Proxy(
  preProxyDefaultValue,
  defaultValueProxyHandler
);

// --- 

function shallowMerge(a, b) {
  // Merge impacts perf: we should like to avoid it if it's possible
  if (a && b) {
    return { ...a, ...b };
  }

  if (a) {
    return a;
  }

  return b;
}

// ---

const hooksFromProps = {
  useButton: 'hook',
  useImage: 'hook',
}
const result = shallowMerge(customStyleHooksContextDefaultValue, hooksFromProps) 
```

⬇️⬇️⬇️

```
result.useButton // string 'hook' ✅
customStyleHooksContextDefaultValue.useFoo // function ✅ 
result.useFoo // undefined 🟥 
```

Reply:
@layershifter Have you done much work with proxies in the past? I'm not sure what the fix would be here if spread breaks the proxy.  Should we do object.assign instead?

Reply:
@GeoffCoxMSFT FYI: I just merged your PR with tests #27086, so we have a good base to test against 🎉 

----

# Option 1

There are few to be aware:
- There is no `isProxy()` in JS, so we don't know with what we are dealing
- Object spread will not work in in this case

To workaround that we can expose `merge()` function on the context:

```diff
// packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts
  useDataGridHeaderCellStyles_unstable: CustomStyleHook;
  useDataGridSelectionCellStyles_unstable: CustomStyleHook;

+ merge: (overrides: CustomStyleHooksContextValue | Partial<CustomStyleHooksContextValue> | undefined) => void;
};
```

And then modify the handler:

```js
const defaultValueProxyHandler: ProxyHandler<CustomStyleHooksContextValue> = {
  get(_target, _prop, _receiver) {
    if (_prop === 'merge') {
      return (overrides: CustomStyleHooksContextValue | Partial<CustomStyleHooksContextValue> | undefined) => {
        if (overrides) {
          Object.assign(preProxyDefaultValue, overrides);
        }
      };
    }

    if (_target[_prop]) {
      return _target[_prop];
    }

    return noop;
  },
};
```

(in this case we mutate `target` object + we are considering `target` as a source of hooks)

Then in `useFluentProvider`:

```diff
// packages/react-components/react-provider/src/components/FluentProvider/useFluentProvider.ts
-  // parentCustomStyleHooks will not be a partial
-  const mergedCustomStyleHooks = shallowMerge(
-    parentCustomStyleHooks,
-    customStyleHooks_unstable,
-  ) as CustomStyleHooksContextValue;
+. parentCustomStyleHooks.merge(customStyleHooks_unstable);

// ...
-  customStyleHooks_unstable: mergedCustomStyleHooks,
+  customStyleHooks_unstable: parentCustomStyleHooks,
```

----

# Option 2

I talked with @ling1726 and his a better idea that does not require `Proxy` and IMO is clearer.

```ts
// The list of hooks is built from the exports from react-components/src/index
type ComponentNames = 'Button' | 'Image' // etc.

type CustomStyleHooksContextValue = {
  (componentName: ComponentNames): CustomStyleHook;
  current: Partial<Record<ComponentNames, CustomStyleHook>>;
};

/**
 * @internal
 */
export const CustomStyleHooksContext = React.createContext<CustomStyleHooksContextValue | undefined>(undefined);

const noop = () => {};

const customStyleHooksContextDefaultValue: CustomStyleHooksContextValue = componentName => {
  return customStyleHooksContextDefaultValue.current[componentName] || noop;
};

customStyleHooksContextDefaultValue.current = {};
```

In this case merging will be just:

```tsx
const parentCustomStyleHooks = useCustomStyleHooks();

parentCustomStyleHooks.current = {
  ...parentCustomStyleHooks,
  ...props.customStyleHooks,
}
```

But it will require changes in components, for example:

```diff
-const { useButtonStyles_unstable: useCustomStyles } = useCustomStyleHooks_unstable();
-useCustomStyles(state);
+const useCustomStyles = useCustomStyleHooks_unstable();
+useCustomStyles('Button')(state)
```

I would say that this approach clearer, but it's up to you to decide 🐱